Symintify pytorch slicing logic

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91340

Differential Revision: [D42398023](https://our.internmc.facebook.com/intern/diff/D42398023)

tugsbayasgalan committed:
Symintify pytorch slicing logic

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42223260).

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

Skylion007 commented on aten/src/ATen/TensorIndexing.h:
```diff
@@ -37,52 +37,48 @@ TORCH_API extern const EllipsisIndexType Ellipsis;
-      step_ = 1;
+      step_ = c10::SymInt(1);
     } else {
       step_ = step_index.value();
```
Now that these are a symints, they aren't trivially copyable so you should move them when you are done with them.
```suggestion
    step_ = std::move(step_index).value();
```

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42223260).

albanD commented on torch/csrc/autograd/python_variable_indexing.h:
```diff
@@ -1,11 +1,94 @@
 #pragma once
 
+#include <c10/core/SymInt.h>
```
Do you really need all these imports?

albanD commented on torch/csrc/autograd/python_variable_indexing.h:
```diff
@@ -1,11 +1,94 @@
+  /* this is harder to get right than you might think */
+
+  // Py_BUILD_ASSERT replaced because it is not available in all versions
+  static_assert(PY_SSIZE_T_MIN + 1 <= -PY_SSIZE_T_MAX, "Build failed");
```
We should use our INDEX_MIN/MAX here instead of the python ssize macros.

tugsbayasgalan replied:
Why does this assert exist? Do we actually need it?

albanD replied:
Just to match the CPython code I'm sure.
But this is most likely because we do assume that these values have this property below and so, since we can do this check for free, we make sure it is true.

albanD replied:
I think we should either keep this or add a comment next to our definition of INDEX_MIN/MAX that we rely on this property

albanD commented on torch/csrc/autograd/python_variable_indexing.h:
```diff
@@ -1,11 +1,94 @@
+
+  auto clip_val = [](Py_ssize_t val) {
+    if (val < c10::SymInt::min_representable_int()) {
+      return c10::SymInt::min_representable_int();
```
Silently truncating sounds wrong here.
If this is user-provided, then I would expect this to be an error.
Otherwise we should not generate such a large number.

albanD replied:
Not sure what you mean? Why would old models have large values and call this code? This is python API code, not the code running the jit serialized model (AFAIK)

albanD commented on torch/csrc/autograd/python_variable_indexing.h:
```diff
@@ -1,11 +1,94 @@
+    }
+  }
+
+  if (torch::is_symint(py::handle(r->start))) {
```
Why does this one needs to be wrapped in a handle but not step in the if statement above?

tugsbayasgalan replied:
my bad. will delete

albanD reviewed (commented):
Sounds mostly ok. Quite a few small details.

Skylion007 commented on aten/src/ATen/TensorIndexing.h:
```diff
@@ -37,52 +37,47 @@ TORCH_API extern const EllipsisIndexType Ellipsis;
-      start_ = step_ < 0 ? INDEX_MAX : 0;
+      start_ = c10::SymInt(step_ < 0 ? INDEX_MAX : 0);
     } else {
       start_ = start_index.value();
```
```suggestion
      start_ = std::move(start_index).value();
```

Skylion007 commented on aten/src/ATen/TensorIndexing.h:
```diff
@@ -37,52 +37,47 @@ TORCH_API extern const EllipsisIndexType Ellipsis;
-      stop_ = step_ < 0 ? INDEX_MIN : INDEX_MAX;
+      stop_ = c10::SymInt(step_ < 0 ? INDEX_MIN : INDEX_MAX);
     } else {
       stop_ = stop_index.value();
```
```suggestion
      stop_ = std::move(stop_index).value();
```
You can also put the std::move after the value() call, but this seems more readable to me. No need to add unnecessary trivial copies with these assignments.

Skylion007 commented on torch/csrc/autograd/python_variable_indexing.h:
```diff
@@ -1,11 +1,94 @@
+    stop_sym = c10::SymInt(stop);
+  }
+
+  return UnpackedSlice{start_sym, stop_sym, step_sym};
```
```suggestion
  return UnpackedSlice{std::move(start_sym), std::move(stop_sym), std::move(step_sym});
```

Skylion007 reviewed (changes requested):
Some small changes that prevent the SymInt from being copied around (and therefore avoiding atomic locks).

voznesenskym commented:
Does this need testing changes? or new tests?

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42223260).

tugsbayasgalan commented:
@voznesenskym pytorch itself has very comprehensive testing suite for slicing logic, so i think as long as CI is green, it should be good.

albanD reviewed (approved):
Sounds good to me.

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42223260).

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_variable_indexing.h:
```diff
@@ -1,11 +1,96 @@
+
+  auto clip_val = [](Py_ssize_t val) {
+    if (val < c10::SymInt::min_representable_int()) {
+      auto r = PyErr_WarnEx(
```
TORCH_WARN("The message");

albanD reviewed (commented):
Small improvements

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42223260).

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42223260).

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42223260).

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42223260).

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

clee2000 commented:
@pytorchbot revert -m "breaking mac builds https://hud.pytorch.org/pytorch/pytorch/commit/8c172fa98a52e95675e9425ac4b23f190f53f9ed https://github.com/pytorch/pytorch/actions/runs/3845932024/jobs/6550654339, marking this as weird because it was merged via codev?" -c weird

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42223260](https://our.internmc.facebook.com/intern/diff/D42223260)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42398023).

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42398023](https://our.internmc.facebook.com/intern/diff/D42398023)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42398023](https://our.internmc.facebook.com/intern/diff/D42398023)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Symintify pytorch slicing logic"


Differential Revision: [D42398023](https://our.internmc.facebook.com/intern/diff/D42398023)

[ghstack-poisoned]

Skylion007 commented on aten/src/ATen/TensorIndexing.h:
```diff
@@ -23,8 +23,8 @@
 
-const int64_t INDEX_MAX = std::numeric_limits<int64_t>::max();
-const int64_t INDEX_MIN = std::numeric_limits<int64_t>::min();
+const int64_t INDEX_MIN = c10::SymInt::min_representable_int();
```
Really tiny nit, this should be constexpr right?

tugsbayasgalan commented:
@pytorchbot merge -f "Landed internally"
